File: SenLib/Sen4/FileFixes/t_item.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen4.FileFixes {
	internal class ItemData {
		public ushort idx;
		public ushort character;
		public string flags;
		public byte[] d1;
		public string name;
		public string desc;
		public byte[] d2;

		internal ItemData(byte[] data) {
			var stream = new DuplicatableByteArrayStream(data);
			idx = stream.ReadUInt16();
			character = stream.ReadUInt16();
			flags = stream.ReadUTF8Nullterm();
			d1 = stream.ReadBytes(0x96);
			name = stream.ReadUTF8Nullterm();
			desc = stream.ReadUTF8Nullterm();
			d2 = stream.ReadBytes(stream.Length - stream.Position);
		}

		internal byte[] ToBinary() {
			MemoryStream ms = new MemoryStream();
			ms.WriteUInt16(idx);
			ms.WriteUInt16(character);
			ms.WriteUTF8Nullterm(flags);
			ms.Write(d1);
			ms.WriteUTF8Nullterm(name);
			ms.WriteUTF8Nullterm(desc);
			ms.Write(d2);
			return ms.CopyToByteArrayAndDispose();
		}
	}

	class t_item : FileMod {
		public string GetDescription() {
			return "Fix incorrect item descriptions and inconsistent formatting.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xe16749c32d8a877cul, 0x4a6181337f7f20b3ul, 0x0e40f379u));
			if (file_en == null) {
				return null;
			}
			var tbl_en = new Tbl(file_en, EndianUtils.Endianness.LittleEndian);

			//for (int i = 0; i < tbl_en.Entries.Count; ++i) {
			//	var e = tbl_en.Entries[i];
			//	if (e.Name == "item" || e.Name == "item_q") {
			//		var m = new ItemData(e.Data);
			//		m.flags = m.flags.Replace("Z", "");
			//		m.desc = m.desc.Replace("\n", " ");
			//		e.Data = m.ToBinary();
			//	}
			//}

			// Tasty Potato Chowder
			{
				int idx = 1468;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.ReplaceSubstring(39, 2, "15", 0, 2); // 2000 HP -> 1500 HP
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Seraph
			{
				int idx = 1211;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				// inconsistent formatting
				m.desc = m.desc.Remove(61, 1);
				m.desc = m.desc.Remove(82, 1);
				m.desc = m.desc.ReplaceSubstring(112, 2, "/", 0, 1);
				m.desc = m.desc.Remove(127, 1);
				m.desc = m.desc.Insert(131, " ");
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Luck
			{
				int idx = 1198;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				// inconsistent formatting
				m.desc = m.desc.Remove(59, 1);
				m.desc = m.desc.Insert(63, " ");
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Hyoushou Gem
			{
				int idx = 1042;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(73, 1); // double space
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Bluster
			{
				int idx = 1118;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(97, 1); // 1 more turns -> 1 more turn
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Aeolus Gem
			{
				int idx = 1134;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(125, 1); // 1 more turns -> 1 more turn
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Cast 1
			{
				int idx = 1144;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(m.desc.Length - 1, 1); // stray ] at the end of the description
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Cast 2
			{
				int idx = 1154;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(m.desc.Length - 1, 1); // stray ] at the end of the description
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Cast 3
			{
				int idx = 1167;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Remove(m.desc.Length - 1, 1); // stray ] at the end of the description
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			// Iron Winged Lion Medal
			{
				int idx = 677;
				var m = new ItemData(tbl_en.Entries[idx].Data);
				m.desc = m.desc.Insert(25, " "); // missing space before parens
				tbl_en.Entries[idx].Data = m.ToBinary();
			}

			Stream result_en = new MemoryStream();
			tbl_en.WriteToStream(result_en, EndianUtils.Endianness.LittleEndian);
			DuplicatableByteArrayStream result_en_b = result_en.CopyToByteArrayStream();
			return new FileModResult[] { new FileModResult("data/text/dat_en/t_item_en.tbl", result_en_b.Duplicate()) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xe16749c32d8a877cul, 0x4a6181337f7f20b3ul, 0x0e40f379u));
			if (file_en == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/text/dat_en/t_item_en.tbl", file_en.Duplicate()) };
		}
	}
}

File: native/sen4/file_fixes/t_item.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen4/tbl.h"
#include "sha1.h"
#include "util/text.h"

namespace SenLib::Sen4::FileFixes::t_item {
std::string_view GetDescription() {
    return "Fix incorrect item descriptions and inconsistent formatting.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    using namespace HyoutaUtils::TextUtils;

    try {
        auto file = getCheckedFile(
            "data/text/dat_en/t_item_en.tbl",
            484211,
            SenPatcher::SHA1FromHexString("4e83e0152b272f6e7739c89a07c1b0c5e2499e8a"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        Tbl tbl_en(bin.data(), bin.size(), HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // for (size_t i = 0; i < tbl_en.Entries.size(); ++i) {
        //     auto& e = tbl_en.Entries[i];
        //     if (e.Name == "item" || e.Name == "item_q") {
        //         ItemData m(e.Data.data(), e.Data.size());
        //         m.flags = Replace(m.flags, "Z", "");
        //         m.desc = Replace(m.desc, "\n", " ");
        //         e.Data = m.ToBinary();
        //     }
        // }

        // Tasty Potato Chowder
        {
            auto& e = tbl_en.Entries[1468];
            ItemData m(e.Data.data(), e.Data.size());
            m.desc = ReplaceSubstring(m.desc, 39, 2, "15", 0, 2); // 2000 HP -> 1500 HP
            e.Data = m.ToBinary();
        }

        // Seraph
        {
            auto& e = tbl_en.Entries[1211];
            ItemData m(e.Data.data(), e.Data.size());
            // inconsistent formatting
            m.desc = Remove(m.desc, 61, 1);
            m.desc = ReplaceSubstring(m.desc, 112, 2, "/", 0, 1);
            m.desc = Remove(m.desc, 127, 1);
            e.Data = m.ToBinary();
        }

        // Luck
        {
            auto& e = tbl_en.Entries[1198];
            ItemData m(e.Data.data(), e.Data.size());
            // inconsistent formatting
            m.desc = Remove(m.desc, 59, 1);
            e.Data = m.ToBinary();
        }

        // Bluster
        {
            auto& e = tbl_en.Entries[1118];
            ItemData m(e.Data.data(), e.Data.size());
            m.desc = Remove(m.desc, 97, 1); // 1 more turns -> 1 more turn
            e.Data = m.ToBinary();
        }

        // Aeolus Gem
        {
            auto& e = tbl_en.Entries[1134];
            ItemData m(e.Data.data(), e.Data.size());
            m.desc = Remove(m.desc, 125, 1); // 1 more turns -> 1 more turn
            e.Data = m.ToBinary();
        }

        // Iron Winged Lion Medal
        {
            auto& e = tbl_en.Entries[677];
            ItemData m(e.Data.data(), e.Data.size());
            m.desc = Insert(m.desc, 25, " "); // missing space before parens
            e.Data = m.ToBinary();
        }

        std::vector<char> result_en_vec;
        MemoryStream result_en(result_en_vec);
        tbl_en.WriteToStream(result_en, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        result.emplace_back(
            std::move(result_en_vec), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen4::FileFixes::t_item
